toolchain/check/handle_modifier.h

```diff
@@ -0,0 +1,37 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_CHECK_HANDLE_MODIFIER_H_
+#define CARBON_TOOLCHAIN_CHECK_HANDLE_MODIFIER_H_
+
+#include "toolchain/check/context.h"
+
+namespace Carbon::Check {
+
+struct DeclModifierKeywords {
+  // At most one of these, and if present it must be first:
+  bool private_ = false;
+  bool protected_ = false;
+
+  // At most one of these allowed for a given declaration:
+  bool abstract_ = false;
+  bool base_ = false;
+  bool default_ = false;
+  bool final_ = false;
+  bool override_ = false;
+  bool virtual_ = false;
+};
+
+// Pops any DeclModifierKeyword parse nodes from `context` and then the
+// introducer node (using `pop_introducer`). Reports a diagnostic if they
+// contain repeated modifiers, modifiers in the incorrect order, or modifiers
+// not in `allowed`. Returns modifiers that were both found and allowed, and the
+// parse node corresponding to the first token of the declaration.
+auto ValidateModifiers(Context& context, DeclModifierKeywords allowed,
+                       std::function<Parse::Node()> pop_introducer)
```

Comment:
Had you considered taking the introducer as NodeKind to pop, instead of a lambda? I think that'd require consider a PopForSoloParseNode version that doesn't template on the enum, which might've been the barrier here.

Reply:
This might be clearer if you look at the other parts of #2987... It was introducing `if constexpr` on the parse kind in order to allow compile-time evaluation of comparisons, versus runtime. In theory `RequireParseKind` alone could take it as a parameter, but then the cases that don't need to put a variable in the stack would need to for the call.

One option would be to hope that the variable and parameter would be optimized away, although I think that could only happen with inlining. The current state is the alternative, just don't pass it as a parameter. Then too, if we switch to CARBON_DCHECK, there's nothing to clean up.

Richard might have a better instinct about how useful an optimization this is, or if we should get rid of this altogether.